### ISSUE TITLE
Can't pickle weight_func in ForecasterAutoregMultiSeries for multiple time series 

### ISSUE THREAD

**Issue by marc-heinl** [skforecast/skforecast#850]
Hello, 

I have problems to save a model when I use the weight_func for multiple series.
Would it be possbile to share a mininmal example of how you can use this for multiple series?

Bellow I have a reproducible example based on the one you shared within your documentation. 
The modelling runs as expected but when I try to save the model with `save_forecaster()` I always run into the following error:

`PicklingError: Can't pickle <function create_weight_func.<locals>.weight_func at 0x7fede452e560>: it's not found as __main__.create_weight_func.<locals>.weight_func`

What I already tried:
- Outsourcing the functions to a separate module
- avoid lambdas --> use hardcoded definition
- register functions als globals()
- generate a class --> run into error that this can't be assigned to weight_func

```python
# Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from lightgbm import LGBMRegressor

from skforecast.plot import set_dark_theme
from skforecast.preprocessing import series_long_to_dict
from skforecast.preprocessing import exog_long_to_dict
from skforecast.ForecasterAutoregMultiSeries import ForecasterAutoregMultiSeries
from skforecast.model_selection_multiseries import backtesting_forecaster_multiseries
from skforecast.model_selection_multiseries import bayesian_search_forecaster_multiseries
from skforecast.utils import save_forecaster, load_forecaster


# Load time series of multiple lengths and exogenous variables
series = pd.read_csv(
    'https://raw.githubusercontent.com/JoaquinAmatRodrigo/skforecast-datasets/main/data/demo_multi_series.csv'
)
exog = pd.read_csv(
    'https://raw.githubusercontent.com/JoaquinAmatRodrigo/skforecast-datasets/main/data/demo_multi_series_exog.csv'
)

series['timestamp'] = pd.to_datetime(series['timestamp'])
exog['timestamp'] = pd.to_datetime(exog['timestamp'])

display(series.head())
print("")
display(exog.head())


# Transform series and exog to dictionaries
series_dict = series_long_to_dict(
    data      = series,
    series_id = 'series_id',
    index     = 'timestamp',
    values    = 'value',
    freq      = 'D'
)

exog_dict = exog_long_to_dict(
    data      = exog,
    series_id = 'series_id',
    index     = 'timestamp',
    freq      = 'D'
)


# Drop some exogenous variables for series 'id_1000' and 'id_1003'
exog_dict['id_1000'] = exog_dict['id_1000'].drop(columns=['air_temperature', 'wind_speed'])
exog_dict['id_1003'] = exog_dict['id_1003'].drop(columns=['cos_day_of_week'])


# Partition data in train and test
end_train = '2016-07-31 23:59:00'

series_dict_train = {k: v.loc[: end_train,] for k, v in series_dict.items()}
exog_dict_train   = {k: v.loc[: end_train,] for k, v in exog_dict.items()}
series_dict_test  = {k: v.loc[end_train:,] for k, v in series_dict.items()}
exog_dict_test    = {k: v.loc[end_train:,] for k, v in exog_dict.items()}


# Plot series
set_dark_theme()
colors = plt.rcParams['axes.prop_cycle'].by_key()['color']
fig, axs = plt.subplots(5, 1, figsize=(8, 4), sharex=True)

for i, s in enumerate(series_dict.values()):
    axs[i].plot(s, label=s.name, color=colors[i])
    axs[i].legend(loc='upper right', fontsize=8)
    axs[i].tick_params(axis='both', labelsize=8)
    axs[i].axvline(pd.to_datetime(end_train) , color='white', linestyle='--', linewidth=1) # End train

fig.suptitle('Series in `series_dict`', fontsize=15)
plt.tight_layout()


# Description of each partition
for k in series_dict.keys():
    print(f"{k}:")
    try:
        print(
            f"\tTrain: len={len(series_dict_train[k])}, {series_dict_train[k].index[0]}"
            f" --- {series_dict_train[k].index[-1]}"
        )
    except:
        print(f"\tTrain: len=0")
    try:
        print(
            f"\tTest : len={len(series_dict_test[k])}, {series_dict_test[k].index[0]}"
            f" --- {series_dict_test[k].index[-1]}"
        )
    except:
        print(f"\tTest : len=0")


series_dates_to_exclude= {
    'id_1000': [
        ('2016-01-01 00:00', '2016-07-31 00:00')  
    ],
    'id_1001': [
        ('2016-07-02 00:00', '2016-07-31 00:00') 
    ],
    'id_1002': [
    ],
    'id_1003': [
        ('2016-01-01 00:00', '2016-07-31 00:00')
    ],
    'id_1004': [
        ('2016-05-02 00:00', '2016-07-31 00:00'),
        ('2016-08-01 00:00', '2016-08-31 00:00')  
    ],
}


# Dictionary to store the weight functions
weight_func_dict = {}


# Function to create a weight function for a specific series based on its date ranges
def create_weight_func(date_ranges):
    def weight_func(index):
        weights = np.ones(len(index))
        for start_date, end_date in date_ranges:
            start_date = pd.Timestamp(start_date)
            end_date = pd.Timestamp(end_date)
            mask = (index >= start_date) & (index <= end_date)
            weights[mask] = 0
        return weights
    return weight_func


# Loop over each series and its date ranges to create weight functions automatically
for series, date_ranges in series_dates_to_exclude.items():
    weight_func_dict[series] = create_weight_func(date_ranges)


# Exogenous variables for each series
for k in series_dict.keys():
    print(f"{k}:")
    try:
        print(f"\t{exog_dict[k].columns.to_list()}")
    except:
        print(f"\tNo exogenous variables")


# Fit forecaster
regressor = LGBMRegressor(random_state=123, verbose=-1, max_depth=5)
forecaster = ForecasterAutoregMultiSeries(
                 regressor          = regressor, 
                 lags               = 14, 
                 encoding           = "ordinal", 
                 dropna_from_series = False,
                 weight_func        = weight_func_dict,
             )

forecaster.fit(series=series_dict_train, exog=exog_dict_train, suppress_warnings=True)

# Predict
predictions = forecaster.predict(steps=5, exog=exog_dict_test, suppress_warnings=True)

metrics_levels, backtest_predictions = backtesting_forecaster_multiseries(
    forecaster            = forecaster,
    series                = series_dict,
    exog                  = exog_dict,
    levels                = None,
    steps                 = 24,
    metric                = "mean_absolute_error",
    add_aggregated_metric = True,
    initial_train_size    = len(series_dict_train["id_1000"]),
    fixed_train_size      = True,
    gap                   = 0,
    allow_incomplete_fold = True,
    refit                 = False,
    n_jobs                ="auto",
    verbose               = True,
    show_progress         = True,
    suppress_warnings     = True
)

display(metrics_levels)
print("")
display(backtest_predictions)

# Plot backtesting predictions
fig, axs = plt.subplots(5, 1, figsize=(8, 4), sharex=True)

for i, s in enumerate(series_dict.keys()):
    axs[i].plot(series_dict[s], label=series_dict[s].name, color=colors[i])
    axs[i].axvline(pd.to_datetime(end_train) , color='white', linestyle='--', linewidth=1) # End train
    try:
        axs[i].plot(backtest_predictions[s], label='prediction', color="white")
    except:
        pass
    axs[i].legend(loc='upper right', fontsize=8)
    axs[i].tick_params(axis='both', labelsize=8)

fig.suptitle('Backtest Predictions', fontsize=15)
plt.tight_layout()

# save skforecast forecaster object 
model_name = type(forecaster.regressor).__name__

forecaster_path = f'../../output/models/forecaster_{model_name}.joblib'
save_forecaster(forecaster, file_name = forecaster_path, save_custom_functions=True)
```

I would appreciate any help.
Thanks in advance and best regards

**Comment by marc-heinl** [comment]
Hi @JavierEscobarOrtiz ,
thanks for the quick response. 
I updatet skforecast to the newest version and tried to execute the example from User Guide for the usage of weight_func_dict for multiple series. I defined the function inside the notebook like in the User Example.
Still I get a pickle error: 
`PicklingError: Can't pickle <function ForecasterRecursiveMultiSeries.create_sample_weights.<locals>.<dictcomp>.<lambda> at 0x7fa046855c60>: it's not found as skforecast.recursive._forecaster_recursive_multiseries.ForecasterRecursiveMultiSeries.create_sample_weights.<locals>.<dictcomp>.<lambda>`

``` python 
# Libraries
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from scipy.stats import skew

from skforecast.datasets import fetch_dataset
from skforecast.recursive import ForecasterRecursive
from skforecast.recursive import ForecasterRecursiveMultiSeries
from skforecast.utils import save_forecaster
from skforecast.utils import load_forecaster


data = fetch_dataset(
    name="h2o", raw=True, kwargs_read_csv={"names": ["y", "date"], "header": 0}
)
data['date'] = pd.to_datetime(data['date'], format='%Y-%m-%d')
data = data.set_index('date')
data = data.asfreq('MS')


def custom_weights_item_1(index):
    """
    Return 0 if index is between 2012-01-01 and 2012-06-01.
    """
    weights = np.where(
        (index >= '2012-01-01') & (index <= '2012-06-01'), 0, 1
    )

    return weights

def custom_weights_item_2(index):
    """
    Return 0 if index is between 2012-04-01 and 2013-01-01.
    """
    weights = np.where(
        (index >= '2012-04-01') & (index <= '2013-01-01'), 0, 1
    )

    return weights

def custom_weights_item_3(index):
    """
    Return 0 if index is between 2012-06-01 and 2013-01-01.
    """
    weights = np.where(
        (index >= '2012-06-01') & (index <= '2013-01-01'), 0, 1
    )

    return weights


class RollingSkewness():
    """
    Custom class to create rolling skewness features.
    """

    def __init__(self, window_sizes, features_names: list = 'rolling_skewness'):
        
        if not isinstance(window_sizes, list):
            window_sizes = [window_sizes]
        self.window_sizes = window_sizes
        self.features_names = features_names

    def transform_batch(self, X: pd.Series) -> pd.DataFrame:
        
        rolling_obj = X.rolling(window=self.window_sizes[0], center=False, closed='left')
        rolling_skewness = rolling_obj.skew()
        rolling_skewness = pd.DataFrame({
                               self.features_names: rolling_skewness
                           }).dropna()

        return rolling_skewness

    def transform(self, X: np.ndarray) -> np.ndarray:
        
        X = X[~np.isnan(X)]
        if len(X) > 0:
            rolling_skewness = np.array([skew(X, bias=False)])
        else:
            rolling_skewness = np.nan
        
        return rolling_skewness

# Create and train ForecasterRecursiveMultiSeries
window_features = RollingSkewness(window_sizes=3)
weight_func_dict = {
    'item_1': custom_weights_item_1,
    'item_2': custom_weights_item_2,
    'item_3': custom_weights_item_3
}

forecaster = ForecasterRecursiveMultiSeries(
                 regressor       = RandomForestRegressor(random_state=123),
                 lags            = 3,
                 window_features = window_features,
                 encoding        = 'ordinal',
                 weight_func     = weight_func_dict
             )

forecaster.fit(series=data)
forecaster

# Save model and custom function
save_forecaster(
    forecaster, 
    file_name = 'forecaster_multiseries_custom_features.joblib', 
    save_custom_functions = True, 
    verbose = False
)


```

**Comment by JavierEscobarOrtiz** [comment]
Hi @marc-heinl 

Thank you for testing the user guide, in your example, you are selecting the wrong data to use, change it to:

```python
# Data download
# ==============================================================================
data = fetch_dataset(name="items_sales")
data.head()
```

However, there is a bug when saving a `ForecasterRecursiveMultiseries` with custom weights. The problem is that if a series is not included in the `weight_func_dict', for instance:

```python
# Not include item_2
weight_func_dict = {
    'item_1': custom_weights_item_1,
    'item_3': custom_weights_item_3
}
```

The Forecaster will create a lambda function that gives a weight of 1 to all observations in this series, and the problem is that Pickle cannot store anonymous functions (lambda).

We will address this issue in the next releases, but as a workaround, you can create a custom function that returns 1 for all elements in the index and assign it to all series that don't require weights.

```python
def custom_weights_all_1(index):
    """
    Return 1 for all elements in the index.
    """
    weights = np.ones(len(index))
    return weights

# item_2 dummy weights
weight_func_dict = {
    'item_1': custom_weights_item_1,
    'item_2': custom_weights_all_1,
    'item_3': custom_weights_item_3
}
```

Please, let me know if it fixes your code.

Thanks!

**Comment by marc-heinl** [comment]
Hi @JavierEscobarOrtiz ,
using custom_weights_all_1() for all series without any specific weights fixed the pickle error. 
Thank you very much!

Best regards!

**Comment by JavierEscobarOrtiz** [comment]
This bug has been fixed in skforecast 0.15.0. Thanks @marc-heinl
